core/src/test/java/org/elasticsearch/action/support/replication/ShardReplicationTests.java

```diff
@@ -304,9 +342,9 @@ public void testRoutingToPrimary() {
 
         TransportReplicationAction<Request, Request, Response>.PrimaryPhase primaryPhase = action.new PrimaryPhase(request, listener);
         assertTrue(primaryPhase.checkBlocks());
-        primaryPhase.routeRequestOrPerformLocally(shardRoutingTable.primaryShard(), shardRoutingTable.shardsIt());
         if (primaryNodeId.equals(clusterService.localNode().id())) {
             logger.info("--> primary is assigned locally, testing for execution");
+            primaryPhase.moveToPrimaryAction(shardRoutingTable.primaryShard());
```

Comment:
Can we somehow split testing of the three phases? For me the transport handler that executes requests and also the AsyncPrimaryAction makes it more difficult to understand which part of TransportReplicationAction is tested.  Or is it very hard to split that?